cpp/src/lance/arrow/dataset.cc

```diff
@@ -316,6 +317,39 @@ ::arrow::Result<std::shared_ptr<UpdaterBuilder>> LanceDataset::NewUpdate(
                                           std::move(new_field));
 }
 
+::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::AddColumn(
+    const std::shared_ptr<::arrow::Field>& field, ::arrow::compute::Expression expression) {
+  auto refs = ::arrow::compute::FieldsInExpression(expression);
+  auto columns = refs | views::transform([](auto& ref) { return ref.ToString(); }) |
+                 to<std::vector<std::string>>;
+  ARROW_ASSIGN_OR_RAISE(auto builder, NewUpdate(field));
+  // Add column
+  // builder->Project(columns);
+  ARROW_ASSIGN_OR_RAISE(auto updater, builder->Finish());
+
+  while (true) {
+    ARROW_ASSIGN_OR_RAISE(auto batch, updater->Next());
+    if (!batch) {
+      break;
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto datum,
+                          ::arrow::compute::ExecuteScalarExpression(expression, *schema(), batch));
```

Comment:
does this bind the expression or is it required to be bound before the AddColumn method is called?

Reply:
the expression is bound when entering this `AddColumn` method.

https://github.com/eto-ai/lance/pull/322/files/088b83b915550453fa1e24fc1e6986c662022a3a#diff-b640faff51137caba09899ac37fb468dd1dbe1cc5e333ce2bb4dd8c43e53a12cR322